### ISSUE TITLE
[BUG] DeepSpeed Inference slowing down text generation (gpt-j)

### ISSUE THREAD

**Issue by rebeccaoskarsson** [microsoft/DeepSpeed#2488]
**Describe the bug**
After implementation of DeepSpeed Inference on locally loaded gpt-j-6B pickled model the text generation latency increases rather than improving it. 
Current results: 
`Vanilla model: P95 latency (ms) - 8582.209633279126; Average latency (ms) - 8565.53 +\- 12.46;
DeepSpeed model: P95 latency (ms) - 8887.944425738533; Average latency (ms) - 8810.54 +\- 65.40;`

**To Reproduce**

1. Download sharded model from Huggingface: `git clone https://huggingface.co/philschmid/gpt-j-6B-fp16-sharded`
2. Create pickled file: 
```from transformers import AutoModelForCausalLM
import torch

model = AutoModelForCausalLM.from_pretrained(
    "gpt-j-6B-fp16-sharded",
    torch_dtype=torch.float16,
)

torch.save(model, "gptj-sharded.pt")
```

3.
Run the following script, updating the path to the pickled model file.
``` import torch
from transformers import AutoTokenizer
import deepspeed
from os import path
from time import perf_counter
import numpy as np
import transformers

# hide generation warnings                                                                                                                                                                                                                                                                 
transformers.logging.set_verbosity_error()

def measure_latency(model, tokenizer, payload, device, generation_args={}):
    input_ids = tokenizer(payload, return_tensors="pt").input_ids.to(device)
    latencies = []
    # warm up                                                                                                                                                                                                                                                                             \
                                                                                                                                                                                                                                                                                           
    for _ in range(2):
        _ =  model.generate(input_ids, **generation_args)
    # Timed run                                                                                                                                                                                                                                                                           \
                                                                                                                                                                                                                                                                                           
    for _ in range(10):
        start_time = perf_counter()
        _ = model.generate(input_ids, **generation_args)
        latency = perf_counter() - start_time
        latencies.append(latency)
    # Compute run statistics                                                                                                                                                                                                                                                              \
                                                                                                                                                                                                                                                                                           
    time_avg_ms = 1000 * np.mean(latencies)
    time_std_ms = 1000 * np.std(latencies)
    time_p95_ms = 1000 * np.percentile(latencies,95)
    return f"P95 latency (ms) - {time_p95_ms}; Average latency (ms) - {time_avg_ms:.2f} +\- {time_std_ms:.2f};", time_p95_ms

# Path to directory with pickled model and config files                                                                                                                                                                                                                                    
model_id = "model/gpt-j-6B-fp16-sharded/"

# load model and tokenizer                                                                                                                                                                                                                                                                 
tokenizer = AutoTokenizer.from_pretrained(model_id)
model = torch.load(path.join(model_id,'gptj-sharded.pt'),map_location='cuda:0')

# init deepspeed inference engine                                                                                                                                                                                                                                                          
ds_model = deepspeed.init_inference(
    model=model,      # Transformers models                                                                                                                                                                                                                                                
    mp_size=1,        # Number of GPU                                                                                                                                                                                                                                                      
    dtype=torch.float16, # dtype of the weights (fp16)                                                                                                                                                                                                                                     
    replace_method="auto", # Lets DS autmatically identify the layer to replace                                                                                                                                                                                                            
    replace_with_kernel_inject=True, # replace the model with the kernel injector                                                                                                                                                                                                          
)
print(f"model is loaded on device {ds_model.module.device}")
from deepspeed.ops.transformer.inference import DeepSpeedTransformerInference

# Ensure DeepSpeed is initialized                                                                                                                                                                                                                                                          
assert isinstance(ds_model.module.transformer.h[0], DeepSpeedTransformerInference) == True, "Model not sucessfully initalized"

# Prepare payload for latency measuring                                                                                                                                                                                                                                                    
payload = (
    "Hello my name is Philipp. I am getting in touch with you because i didn't get a response from you. What do I need to do to get my new card which I have requested 2 weeks ago? Please help me and answer this email in the next 7 days. Best regards and have a nice weekend but it"
    * 2
)
print(f'Payload sequence length is: {len(tokenizer(payload)["input_ids"])}')

# generation arguments                                                                                                                                                                                                                                                                     
generation_args = dict(do_sample=False, num_beams=1, min_length=128, max_new_tokens=128)
vanilla_results = measure_latency(model=model,tokenizer=tokenizer,payload=payload,generation_args=generation_args, device=model.device)
ds_results = measure_latency(model=ds_model, tokenizer=tokenizer,payload=payload, generation_args=generation_args, device=ds_model.module.device)

print(f"Vanilla model: {vanilla_results[0]}")
print(f"DeepSpeed model: {ds_results[0]}")
```


**Expected behavior**
Improvement of latency with about 2 seconds for the deepspeed model, at least. 

**ds_report output**
```--------------------------------------------------
DeepSpeed C++/CUDA extension op report
--------------------------------------------------
NOTE: Ops not installed will be just-in-time (JIT) compiled at
      runtime if needed. Op compatibility means that your system
      meet the required dependencies to JIT install the op.
--------------------------------------------------
JIT compiled ops requires ninja
ninja .................. [OKAY]
--------------------------------------------------
op name ................ installed .. compatible
--------------------------------------------------
cpu_adam ............... [NO] ....... [OKAY]
cpu_adagrad ............ [NO] ....... [OKAY]
fused_adam ............. [NO] ....... [OKAY]
fused_lamb ............. [NO] ....... [OKAY]
 [WARNING]  please install triton==1.0.0 if you want to use sparse attention
sparse_attn ............ [NO] ....... [NO]
transformer ............ [NO] ....... [OKAY]
stochastic_transformer . [NO] ....... [OKAY]
 [WARNING]  async_io requires the dev libaio .so object and headers but these were not found.
 [WARNING]  async_io: please install the libaio-dev package with apt
 [WARNING]  If libaio is already installed (perhaps from source), try setting the CFLAGS and LDFLAGS environment variables to where it can be found.
async_io ............... [NO] ....... [NO]
utils .................. [NO] ....... [OKAY]
quantizer .............. [NO] ....... [OKAY]
transformer_inference .. [NO] ....... [OKAY]
--------------------------------------------------
DeepSpeed general environment info:
torch install path ............... ['/home/rebeccaoskarsson/.local/lib/python3.9/site-packages/torch']
torch version .................... 1.11.0+cu115
torch cuda version ............... 11.5
torch hip version ................ None
nvcc version ..................... 11.5
deepspeed install path ........... ['/home/rebeccaoskarsson/.local/lib/python3.9/site-packages/deepspeed']
deepspeed info ................... 0.7.4+fffca7df, fffca7df, ds-inference/support-large-token-length
deepspeed wheel compiled w. ...... torch 1.11, cuda 11.5
```

**Screenshots**

**System info (please complete the following information):**
 - OS: Debian GNU/Linux 11 (bullseye)
 - GPU count and types: One machine with one Nvidia Tesla T4
 - DeepSpeed: deepspeed @ git+https://github.com/microsoft/DeepSpeed.git@fffca7dfd84fafc6a6dd04ded7e7132b08f2c737
 - Hugging Face Transformers/Accelerate/etc. versions: transformers==4.21.2
 - Python version: Python 3.9.2

**Additional context**
Followed a [tutorial by Phillip Schmid](https://www.philschmid.de/gptj-deepspeed-inference) and have been in touch with him. We agreed to report this as a bug, as we could not figure out the problem.

**Comment by RezaYazdaniAminabadi** [comment]
Hi @rebeccaoskarsson 

Thanks for trying out DS-Inference. I am gonna take a look at this now, and try to repro on my side. I will update you as I find out more about the issue?
Thanks,
Reza

**Comment by RezaYazdaniAminabadi** [comment]
So, I think I found where the issue is here. 
Since you have done the `init_inference` before running inference with the model (called as Vanilla model), it has already been injected with the DeepSpeed kernels (note that the injection is done in place). So, you were basically running with deepspeed in both cases. I also ran this on my side on a T4 machine and I got the following results for DeepSpeed and Baseline:

```
DeepSpeed: 5312.46 ms
Baseline: 7348.63 ms
Speedup: 1.39x
```
I am using the latest DeepSpeed branch and Pytorch 1.11.0+cu113 and transformers 4.24.0

Thanks,
Reza
